File: 0152-maximum-product-subarray/0152-maximum-product-subarray.java
class Solution {
    public int maxProduct(int[] nums) {
        if (nums == null | nums.length == 0) {
            return 0;
        }

        int maxProduct = nums[0];
        int minProduct = nums[0];
        int result = nums[0];

        for (int i = 1; i < nums.length; i++) {
            if (nums[i] >= 0) {
                maxProduct = Math.max(nums[i], nums[i] * maxProduct);
                minProduct = Math.min(nums[i], nums[i] * minProduct);
            } else {
                int temp = maxProduct;
                maxProduct = Math.max(nums[i], nums[i] * minProduct);
                minProduct = Math.min(nums[i], nums[i] * temp);
            }

            result = Math.max(result, maxProduct);
        }

        return result;
    }
}

File: 0152-maximum-product-subarray/0152-maximum-product-subarray.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        if len(nums) == 0:
            return 0
        
        maxProduct = nums[0]
        minProduct = nums[0]
        result = nums[0]

        for num in nums[1:]:
            if num >= 0:
                maxProduct = max(num, num * maxProduct)
                minProduct = min(num, num * minProduct)
            else:
                temp = maxProduct
                maxProduct = max(num, num * minProduct)
                minProduct = min(num, num * temp)
            
            result = max(result, maxProduct)

        return result 